test/lib/jdk/test/whitebox/WhiteBox.java

```diff
@@ -556,6 +560,53 @@ public void clearInlineCaches(boolean preserve_static_stubs) {
   // Force Full GC
   public native void fullGC();
 
+  // Infrastructure for waitForReferenceProcessing()
+  private static volatile Method waitForReferenceProcessingMethod = null;
+
+  private static Method getWaitForReferenceProcessingMethod() {
+    Method wfrp = waitForReferenceProcessingMethod;
+    if (wfrp == null) {
+      try {
+        wfrp = Reference.class.getDeclaredMethod("waitForReferenceProcessing");
+        wfrp.setAccessible(true);
+        assert wfrp.getReturnType() == Boolean.class;
+        Class<?>[] ev = wfrp.getExceptionTypes();
+        assert ev.length == 1;
+        assert ev[0] == InterruptedException.class;
+        waitForReferenceProcessingMethod = wfrp;
+      } catch (InaccessibleObjectException e) {
+        throw new RuntimeException("Need to add @modules java.base/java.lang.ref:open to test?", e);
+      } catch (NoSuchMethodException e) {
+        throw new RuntimeException(e);
+      }
+    }
+    return wfrp;
+  }
+
+  /**
+   * Wait for reference processing, via Reference.waitForReferenceProcessing().
+   * Callers of this method will need the
+   * @modules java.base/java.lang.ref:open
+   * jtreg tag.
+   *
+   * This method should usually be called after a call to WhiteBox.fullGC().
+   */
+  public boolean waitForReferenceProcessing() throws InterruptedException {
+    try {
+      Method wfrp = getWaitForReferenceProcessingMethod();
+      return (Boolean) wfrp.invoke(null);
+    } catch (IllegalAccessException t) {
+      throw new RuntimeException("Shouldn't happen, we call setAccessible()", t);
+    } catch (InvocationTargetException e) {
+      Throwable cause = e.getCause();
+      if (cause instanceof InterruptedException) {
+        throw (InterruptedException) cause;
+      } else {
+        throw new RuntimeException(e);
```

Comment:
I still think it would be better to finish unwrapping e, rather than further
wrapping it. We know the only checked exception is InterruptedException,
because we checked that when looking up the method. That leaves either
RuntimeException or Error. I think it is generally inappropriate to wrap
Errors, and we don't really need to additionally wrap the RuntimeException
case. But I'm going to approve this change anyway, and leave that question to
real Java programmers.

Reply:
For production code, I might do more.
But for a test library (where, IMO, the exception handling path won't see much action), doing enough to get the information into the test log seems sufficient to me.